### ISSUE TITLE
UnusedPrivateMember false positives

### ISSUE THREAD

**Issue by MyDogTom** [arturbosch/detekt#812]
version: 1.0.0.RC6-4
**Case 1**

```kotlin
@ConsumerScope
@Component(dependencies = arrayOf(OneComponent::class))
interface AnotherComponent {
    fun inject(myActivity: MyActivity)
}
```
Rule complains about `myActivity`.

```kotlin
interface UserPlugin {
    fun plug(application: Application)
    fun unplug()
}
```

Rule complains about `application`

**Case 2**

```kotlin
    override fun funA() {
        objectA.resolve(valA, object : MyCallback {
            override fun onResolveFailed(throwable: Throwable) {
                errorMessage.visibility = View.VISIBLE
            }
        })
    }
```

rule complains about `throwable`

**Case 3**
```kotlin 
class A {
    companion object {
        private const val MY_CONST = 42
    }

    fun a() {
        Compleatable.timer(MY_CONST.toLong(), TimeUnit.MILLISECONDS)
              .subscribe()
    }
}
```

rule complains about `MY_CONST`

PS. I have many more violations in my codebase. I just randomly picked few of them, but most likely all of them a similar.

**Comment by vanniktech** [comment]
I've got a few more:

### Case 4

Function being flagged

```
UnusedPrivateMember - [integrationTest] at gradle-dependency-graph-generator-plugin/src/test/java/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorPluginTest.kt:61:3
```

(https://github.com/vanniktech/gradle-dependency-graph-generator-plugin/blob/167e6401ba7245e0f0a724851d5f18decca285b1/src/test/java/com/vanniktech/dependency/graph/generator/DependencyGraphGeneratorPluginTest.kt#L61)

### Another Case 2
`b` is unused. Gets flagged even twice. (Same as Case 2):

```kotlin
internal object NullOutputStream : OutputStream() {
  override fun write(b: Int) {
    // No-op.
  }
}
```

Parameters in overridden methods should not be taken into account.


### Case 5

Extension function gets flagged:

```
UnusedPrivateMember - [getDisplayName] at gradle-dependency-graph-generator-plugin/src/main/kotlin/com/vanniktech/dependency/graph/generator/DotGenerator.kt:82:3
```

https://github.com/vanniktech/gradle-dependency-graph-generator-plugin/blob/167e6401ba7245e0f0a724851d5f18decca285b1/src/main/kotlin/com/vanniktech/dependency/graph/generator/DotGenerator.kt#L82

-----

Is it possible to run new rules against open source projects to verify that there are no false positives?

**Comment by arturbosch** [comment]
That would be optimal :D, currently we leave them turned off and let the early birds test them :D.

**Comment by wtanaka** [comment]
```
val test = object : Iterator<String?> {

    var mutatable: String? = null

    private fun preCall() {
        mutatable = "done"
    }

    override fun next(): String? {
        preCall()
        return mutatable
    }

    override fun hasNext(): Boolean = true
}

fun main(args: Array<String>) {
    println(test.next())
}
```
